proto: make the tonic/transport feature optional

bouk opened the pull request:
I tried compiling a wasm client for opentelemetry-rust with a custom `HttpClient`, but `opentelemetry-proto` depends on `tonic` which has a dependency on tokio. This can't be compiled for wasm, of course. Here I'm adding a feature `tonic-transport` to `opentelemetry-proto` which will only enable the client and server implementations if `tonic-transport` is enabled.

This should make it possible to compile opentelemetry-proto for wasm.

Related: https://github.com/open-telemetry/opentelemetry-rust/issues/696

bouk commented on opentelemetry-otlp/Cargo.toml:
```diff
@@ -68,7 +68,7 @@ serialize = ["serde"]
 
 # grpc using tonic
-grpc-tonic = ["tonic", "prost", "http", "tokio", "opentelemetry-proto/gen-tonic"]
+grpc-tonic = ["tonic", "prost", "http", "tokio", "opentelemetry-proto/gen-tonic", "opentelemetry-proto/tonic-transport"]
```
No, because `gen-tonic` is needed by `opentelemetry-otlp/http-proto` to encode the protobuf messages. `opentelemetry-otlp/http-proto` including `tonic/transport`  by default is what I'm trying to avoid here.

djc replied:
Yeah, but you want to change the defaults for a relatively niche use case. I'm suggesting we shouldn't do that, and merely make it possible to address your use case, without making it the default.

bouk replied:
Right I see what you mean, I can add `tonic-transport` to the `opentelemetry-proto` defaults so it's a backwards-compatible change :+1:

djc replied:
Yup, that sounds good!

bouk commented:
@djc I've moved things around and now it's much cleaner:

* Introduced a `gen-tonic-messages` feature to `opentelemetry-proto` that only generates the messages
* `opentelemetry-proto/gen-tonic` now includes `tonic/transport` and `gen-tonic-messages`
* `http-proto` now depends on `opentelemetry-proto/gen-tonic-messages` instead of `gen-tonic`

This is backwards-compatible and accomplishes what I was trying to do

djc commented on opentelemetry-otlp/Cargo.toml:
```diff
@@ -78,7 +78,7 @@ openssl = ["grpcio/openssl"]
 
 # http binary
-http-proto = ["prost", "opentelemetry-http", "opentelemetry-proto/gen-tonic", "http", "trace"]
+http-proto = ["prost", "opentelemetry-http", "opentelemetry-proto/gen-tonic-messages", "http", "trace"]
```
With the changes you've made in `-proto`, I think all mentions of `gen-tonic` in `-otlp` should start using `gen-tonic-messages`.

bouk replied:
Which other mentions do you mean? `grpc-tonic` needs `gen-tonic` because it uses the tonic client.

djc reviewed (approved):
Okay, this looks good to me -- just needs a rebase now.

(Sorry for the slow feedback.)

bouk commented:
@djc done!

hdost reviewed (approved):
Looks good. Thanks for the contribution. 👍🏼

djc commented:
Not sure what's up with the beta test in CI here. Locally if I run `cargo +beta test` in opentelemetry-proto with the default features it passes. @bouk can you have a look whether you can reproduce/if this is related to your changes?

bouk commented:
@djc it seems I messed up the rebase before, I think it's fixed now

djc commented:
Thanks!
